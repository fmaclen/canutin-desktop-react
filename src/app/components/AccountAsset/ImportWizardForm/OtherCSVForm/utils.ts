import { parse, format } from 'date-fns';

import { CanutinJsonTransactionType } from '@appTypes/canutin';
import { Account } from '@database/entities';
import { getBalanceGroupByAccountType } from '@database/helpers';
import { CanutinJsonType, UpdatedAccount } from '@appTypes/canutin';

import {
  SupportedDateFormatType,
} from './otherCsvConstants';
import { OtherCSVFormSubmit } from './index';

export const getTransactionsForOneAccount = (
  csvData: { [columnName: string]: string }[],
  descriptionColumn: string,
  dateColumn: string,
  dateFormat: SupportedDateFormatType,
  amountColumn: string,
  categoryColumn: string | null,
  categoryValues?: { [categoryColumnValue: string]: string }
) => {
  const transactions = csvData.map((rowData: { [x: string]: any }) => ({
    description: rowData[descriptionColumn],
    date: format(parse(rowData[dateColumn], dateFormat, new Date()), 'MM/dd/yyyy'),
    amount: Number(rowData[amountColumn]),
    excludeFromTotals: true,
    category:
      categoryColumn && categoryValues ? categoryValues[rowData[categoryColumn]] : 'Uncategorized',
  }));

  return transactions;
};

export const getTransactionsForAccountColumn = (
  accountColumn: string,
  accounts: { [accountColumnValue: string]: string } = {},
  csvData: { [columnName: string]: string }[],
  descriptionColumn: string,
  dateColumn: string,
  dateFormat: SupportedDateFormatType,
  amountColumn: string,
  categoryColumn: string | null,
  categoryValues?: { [categoryColumnValue: string]: string }
) => {
  const cFile: CanutinJsonType = { accounts: [] };
  Object.keys(accounts).forEach(accountColumnName => {
    const transactions: CanutinJsonTransactionType[] = [];

    csvData.forEach((rowData: { [x: string]: any }) => {
      if (rowData[accountColumn] === accountColumnName) {
        transactions.push({
          description: rowData[descriptionColumn],
          date: format(parse(rowData[dateColumn], dateFormat, new Date()), 'MM/dd/yyyy'),
          amount: Number(rowData[amountColumn]),
          excludeFromTotals: true,
          category:
            categoryColumn && categoryValues
              ? categoryValues[rowData[categoryColumn]]
              : 'Uncategorized',
        });
      }
    });

    const account = {
      name: accountColumnName,
      balanceGroup: getBalanceGroupByAccountType(accounts[accountColumnName]),
      accountType: accounts[accountColumnName],
      autoCalculate: true,
      transactions,
    };

    cFile.accounts.push(account);
  });

  return cFile;
};

export const getUpdatedTransactionsForExistingAccounts = (
  canutinAccounts: Account[] | null,
  accountColumn: string,
  csvData: { [columnName: string]: string }[],
  descriptionColumn: string,
  dateColumn: string,
  dateFormat: SupportedDateFormatType,
  amountColumn: string,
  categoryColumn: string | null,
  categoryValues?: { [categoryColumnValue: string]: string }
) => {
  const updatedAccounts: UpdatedAccount[] = [];
  csvData.forEach(rowData => {
    const updateAccount = canutinAccounts?.find(account => account.name === rowData[accountColumn]);
    console.log(canutinAccounts);
    console.log(rowData[accountColumn]);
    if (updateAccount) {
      const accountAlreadyCreated = updatedAccounts.find(
        account => account.id === updateAccount.id
      );

      if (!accountAlreadyCreated) {
        const transactions: CanutinJsonTransactionType[] = [
          {
            description: rowData[descriptionColumn],
            date: format(parse(rowData[dateColumn], dateFormat, new Date()), 'MM/dd/yyyy'),
            amount: Number(rowData[amountColumn]),
            excludeFromTotals: true,
            category:
              categoryColumn && categoryValues
                ? categoryValues[rowData[categoryColumn]]
                : 'Uncategorized',
          },
        ];
        updatedAccounts.push({ id: updateAccount.id, transactions });
      }

      if (accountAlreadyCreated) {
        accountAlreadyCreated.transactions.push({
          description: rowData[descriptionColumn],
          date: format(parse(rowData[dateColumn], dateFormat, new Date()), 'MM/dd/yyyy'),
          amount: Number(rowData[amountColumn]),
          excludeFromTotals: true,
          category:
            categoryColumn && categoryValues
              ? categoryValues[rowData[categoryColumn]]
              : 'Uncategorized',
        });
      }
    }
  });

  return updatedAccounts;
};

export const formToCantuinJsonFile = (
  formData: OtherCSVFormSubmit,
  csvData: any,
  canutinAccounts: Account[] | null
) => {
  const {
    accountColumn,
    amountColumn,
    categoryColumn,
    dateColumn,
    dateFormat,
    descriptionColumn,
    categories,
    accounts,
  } = formData;
  let canutinFile: CanutinJsonType = { accounts: [] };

  // New account or update account
  if (formData.account) {
    const { name, accountType, autoCalculate, balance, institution } = formData.account;
    const transactions = getTransactionsForOneAccount(
      csvData,
      descriptionColumn,
      dateColumn,
      dateFormat,
      amountColumn,
      categoryColumn,
      categories
    );
    canutinFile = {
      accounts: [
        {
          name,
          balanceGroup: getBalanceGroupByAccountType(accountType),
          accountType,
          institution,
          autoCalculate,
          balance,
          transactions,
        },
      ],
    };

    return { canutinFile };
  }

  // Multiple new accounts
  if (accountColumn) {
    canutinFile = getTransactionsForAccountColumn(
      accountColumn,
      accounts,
      csvData,
      descriptionColumn,
      dateColumn,
      dateFormat,
      amountColumn,
      categoryColumn,
      categories
    );
    const updatedAccounts = getUpdatedTransactionsForExistingAccounts(
      canutinAccounts,
      accountColumn,
      csvData,
      descriptionColumn,
      dateColumn,
      dateFormat,
      amountColumn,
      categoryColumn,
      categories
    );

    return { canutinFile, updatedAccounts };
  }

  return canutinFile;
};
