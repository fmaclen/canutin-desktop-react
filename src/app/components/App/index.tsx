import React, { useEffect, useState, useContext } from 'react';
import { Switch, Route, BrowserRouter, Redirect } from 'react-router-dom';
import { ipcRenderer, IpcRendererEvent } from 'electron';
import styled from 'styled-components';

import TitleBar from '@components/common/TitleBar';
import StatusBar from '@components/common/StatusBar';
import SideBar from '@components/common/SideBar';
import { AppContext } from '@app/context/appContext';
import { requestSync, getLinkSummary } from '@app/data/canutinLink.api';

import Setup from '@pages/Setup';

import { routesConfig, RouteConfigProps, routesPaths } from '@routes';
import { DATABASE_CONNECTED, DATABASE_NOT_DETECTED } from '@constants';
import { DB_GET_ACCOUNTS_ACK, DB_GET_ASSETS_ACK } from '@constants/events';
import AssetIpc from '@app/data/asset.ipc';
import AccountIpc from '@app/data/account.ipc';
import { Account, Asset } from '@database/entities';

import GlobalStyle from '@app/styles/global';
import { container } from './styles';

const Container = styled.div`
  ${container}
`;

const App = () => {
  const {
    isLoading,
    setIsLoading,
    isAppInitialized,
    setIsAppInitialized,
    setFilePath,
    isDbEmpty,
    setIsDbEmpty,
    setLinkAccount,
    linkAccount,
  } = useContext(AppContext);
  const [accounts, setAccounts] = useState<Account[] | null>(null);
  const [assets, setAssets] = useState<Asset[] | null>(null);

  useEffect(() => {
    ipcRenderer.on(DB_GET_ACCOUNTS_ACK, (_: IpcRendererEvent, accounts: Account[]) => {
      setAccounts(accounts);
    });

    ipcRenderer.on(DB_GET_ASSETS_ACK, (_: IpcRendererEvent, assets: Asset[]) => {
      setAssets(assets);
    });

    ipcRenderer.on(DATABASE_CONNECTED, (_, filePath) => {
      setIsLoading(false);
      setIsAppInitialized(true);
      setFilePath(filePath?.filePath);
      if (filePath) {
        AssetIpc.getAssets();
        AccountIpc.getAccounts();
      }
    });

    ipcRenderer.on(DATABASE_NOT_DETECTED, () => {
      setIsLoading(false);
      setIsAppInitialized(false);
    });

    return () => {
      ipcRenderer.removeAllListeners(DATABASE_NOT_DETECTED);
      ipcRenderer.removeAllListeners(DATABASE_CONNECTED);
    };
  }, []);

  useEffect(() => {
    if (
      Array.isArray(assets) &&
      assets.length === 0 &&
      Array.isArray(accounts) &&
      accounts.length === 0
    ) {
      setIsDbEmpty(true);
    } else {
      setIsDbEmpty(false);

      const handleSync = async () => {
        const summary = await getLinkSummary();
        summary && setLinkAccount(summary);
      };

      handleSync();
    }
  }, [assets, accounts]);

  useEffect(() => {
    if (linkAccount?.isSyncing) {
      const handleSync = async () => {
        const summary = await getLinkSummary();
        summary && setLinkAccount(summary);

        await requestSync();
        setLinkAccount({ ...linkAccount, isSyncing: false });
      };

      handleSync();
    }
  }, [linkAccount]);

  return (
    <>
      <GlobalStyle />
      <BrowserRouter>
        <Container>
          {!isLoading && isAppInitialized && (
            <>
              <TitleBar />
              <SideBar />
              <Switch>
                {isDbEmpty && (
                  <Redirect
                    exact
                    from={routesPaths.bigpicture}
                    to={routesPaths.addAccountOrAsset}
                  />
                )}
                {routesConfig.map(({ path, component, exact }: RouteConfigProps, index) => (
                  <Route key={index} exact={exact} path={path}>
                    {component}
                  </Route>
                ))}
              </Switch>
            </>
          )}

          {!isAppInitialized && (
            <>
              <TitleBar />
              <Setup />
            </>
          )}
          <StatusBar />
        </Container>
      </BrowserRouter>
    </>
  );
};

export default App;
