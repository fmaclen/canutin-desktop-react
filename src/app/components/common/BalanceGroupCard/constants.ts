export enum BalanceGroupCardTypeEnum {
  CASH,
  DEBT,
  INVESTMENTS,
  OTHER_ASSETS,
  NET_WORTH,
}

export const getBalanceGroupCardTitle = {
  [BalanceGroupCardTypeEnum.CASH]: 'Cash',
  [BalanceGroupCardTypeEnum.DEBT]: 'Debt',
  [BalanceGroupCardTypeEnum.INVESTMENTS]: 'Investments',
  [BalanceGroupCardTypeEnum.OTHER_ASSETS]: 'Other assets',
  [BalanceGroupCardTypeEnum.NET_WORTH]: 'Net worth',
};
