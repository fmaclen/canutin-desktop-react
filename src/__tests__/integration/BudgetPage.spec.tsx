import { screen, waitFor } from '@testing-library/react';
import { ipcRenderer, IpcRendererEvent } from 'electron';
import { mocked } from 'ts-jest/utils';
import userEvent from '@testing-library/user-event';
import { endOfDay, format, startOfDay, subMonths } from 'date-fns';

import { render } from '@tests/utils';
import App from '@components/App';
import {
  DB_GET_ACCOUNTS_ACK,
  DB_GET_BUDGETS_ACK,
  DB_GET_TRANSACTION_CATEGORY_ACK,
  FILTER_TRANSACTIONS_ACK,
} from '@constants/events';
import { DATABASE_CONNECTED } from '@constants';
import { AppCtxProvider } from '@app/context/appContext';
import { EntitiesProvider } from '@app/context/entitiesContext';
import { TransactionsProvider } from '@app/context/transactionsContext';
import { seedAccounts } from '@tests/factories/seededEntitiesFactory';
import { autoBudgetCategoriesBuilder } from '@tests/factories/autoBudgetCategoriesFactory';
import {
  accountCheckingDetails,
  accountCreditCardDetails,
  accountSavingsDetails,
} from '@database/seed/demoData/accounts';
import {
  accountCheckingTransactionSet,
  accountSavingsTransactionSet,
  accountCreditCardTransactionSet,
} from '@database/seed/demoData/transactions';
import { filters } from '@app/constants/filters';
import { dateInUTC } from '@app/utils/date.utils';
import mapCategories from '@database/helpers/importResources/mapCategories';
import BudgetIpc from '@app/data/budget.ipc';

const initAppWithContexts = () => {
  render(
    <AppCtxProvider>
      <EntitiesProvider>
        <TransactionsProvider>
          <App />
        </TransactionsProvider>
      </EntitiesProvider>
    </AppCtxProvider>
  );
};

describe('Budget tests', () => {
  const minimumAccount = [{ ...accountCheckingDetails, transactions: [] }];

  test("Sidebar link can't be clicked if no accounts or assets are present", async () => {
    mocked(ipcRenderer).on.mockImplementation((event, callback) => {
      if (event === DATABASE_CONNECTED) {
        callback((event as unknown) as IpcRendererEvent, {
          filePath: 'testFilePath',
        });
      }

      return ipcRenderer;
    });

    initAppWithContexts();
    const budgetSidebarLink = screen.getByTestId('sidebar-budget');
    expect(budgetSidebarLink).toHaveAttribute('disabled');

    userEvent.click(budgetSidebarLink);
    expect(budgetSidebarLink).not.toHaveAttribute('active', '1');
  });

  test('Budget page displays an empty view when no enough data is available', async () => {
    mocked(ipcRenderer).on.mockImplementation((event, callback) => {
      if (event === DATABASE_CONNECTED) {
        callback((event as unknown) as IpcRendererEvent, {
          filePath: 'testFilePath',
        });
      }

      if (event === DB_GET_ACCOUNTS_ACK) {
        callback((event as unknown) as IpcRendererEvent, minimumAccount);
      }

      if (event === FILTER_TRANSACTIONS_ACK) {
        callback((event as unknown) as IpcRendererEvent, { transactions: [] });
      }

      return ipcRenderer;
    });

    initAppWithContexts();
    const budgetSidebarLink = screen.getByTestId('sidebar-budget');
    expect(budgetSidebarLink).not.toHaveAttribute('disabled');

    userEvent.click(budgetSidebarLink);
    expect(budgetSidebarLink).toHaveAttribute('active', '1');
    expect(screen.getByText('Auto-budget')).toBeVisible();
    expect(
      screen.getByText(
        'Need at least 2 months of transactions for budgets to be displayed when auto-budget is enabled'
      )
    ).toBeVisible();

    const periodChooser = screen.getByText(format(new Date(), 'MMMM yyyy'));
    expect(periodChooser).toBeVisible();

    userEvent.click(periodChooser);
    expect(screen.getByText(format(subMonths(new Date(), 6), 'MMMM yyyy'))).toBeVisible();
    expect(screen.getByText(format(subMonths(new Date(), 12), 'MMMM yyyy'))).toBeVisible();
  });

  test('Budget page displays the correct data when auto-budget is enabled', async () => {
    let oneMonthOfTransactions = accountCheckingTransactionSet()
      .filter(transaction => {
        // Mimic the logic in `TransactionRepository.getFilterTransactions()`
        const dateFrom = dateInUTC(startOfDay(filters[0].dateFrom));
        const dateTo = dateInUTC(endOfDay(filters[0].dateTo));
        return transaction.date >= dateFrom && transaction.date <= dateTo;
      })
      .map(transaction => {
        return {
          ...transaction,
          account: { ...accountCheckingDetails },
          category: { name: mapCategories(transaction.categoryName) },
        };
      });

    accountSavingsTransactionSet()
      .filter(transaction => {
        const dateFrom = dateInUTC(startOfDay(filters[0].dateFrom));
        const dateTo = dateInUTC(endOfDay(filters[0].dateTo));
        return transaction.date >= dateFrom && transaction.date <= dateTo;
      })
      .forEach(transaction => {
        oneMonthOfTransactions.push({
          ...transaction,
          account: { ...accountSavingsDetails },
          category: { name: mapCategories(transaction.categoryName) },
        });
      });

    accountCreditCardTransactionSet()
      .filter(transaction => {
        const dateFrom = dateInUTC(startOfDay(filters[0].dateFrom));
        const dateTo = dateInUTC(endOfDay(filters[0].dateTo));
        return transaction.date >= dateFrom && transaction.date <= dateTo;
      })
      .forEach(transaction => {
        oneMonthOfTransactions.push({
          ...transaction,
          account: { ...accountCreditCardDetails },
          category: { name: mapCategories(transaction.categoryName) },
        });
      });

    // Out of budget transaction
    oneMonthOfTransactions.push({
      description: 'Better Purchase Electronics Co.',
      amount: -99.99,
      date: new Date(),
      categoryName: 'Uncategorized',
      excludeFromTotals: false,
      account: { ...accountCreditCardDetails },
      category: { name: mapCategories('Uncategorized') },
    });

    const {
      needsCategories,
      wantsCategories,
      transactionsWithCategories,
    } = autoBudgetCategoriesBuilder(oneMonthOfTransactions);
    oneMonthOfTransactions = transactionsWithCategories;

    mocked(ipcRenderer).on.mockImplementation((event, callback) => {
      if (event === DATABASE_CONNECTED) {
        callback((event as unknown) as IpcRendererEvent, {
          filePath: 'testFilePath',
        });
      }

      if (event === DB_GET_ACCOUNTS_ACK) {
        callback((event as unknown) as IpcRendererEvent, seedAccounts);
      }

      if (event === FILTER_TRANSACTIONS_ACK) {
        callback((event as unknown) as IpcRendererEvent, {
          transactions: oneMonthOfTransactions,
        });
      }

      if (event === DB_GET_TRANSACTION_CATEGORY_ACK) {
        needsCategories.forEach(category => {
          callback((event as unknown) as IpcRendererEvent, category);
        });
        wantsCategories.forEach(category => {
          callback((event as unknown) as IpcRendererEvent, category);
        });
      }

      if (event === DB_GET_BUDGETS_ACK) {
        callback((event as unknown) as IpcRendererEvent, []);
      }

      return ipcRenderer;
    });

    initAppWithContexts();
    userEvent.click(screen.getByTestId('sidebar-big-picture')); // Resets path back to the default
    const budgetSidebarLink = screen.getByTestId('sidebar-budget');
    expect(budgetSidebarLink).toHaveAttribute('active', '0');
    expect(budgetSidebarLink).not.toHaveAttribute('disabled');

    userEvent.click(budgetSidebarLink);
    expect(budgetSidebarLink).toHaveAttribute('active', '1');

    const budgetBarIncome = screen.getByTestId('budget-bar-income');
    expect(budgetBarIncome).toHaveTextContent('Income');
    expect(budgetBarIncome).toHaveTextContent('$7,577');
    expect(budgetBarIncome).toHaveTextContent('$8,050 (106%)');

    const budgetBarExpenses = screen.getByTestId('budget-bar-expenses');
    expect(budgetBarExpenses).toHaveTextContent('Expenses');
    expect(budgetBarExpenses).toHaveTextContent('-$6,061');
    expect(budgetBarExpenses).toHaveTextContent('-$7,922 (131%)');

    const budgetBarSavings = screen.getByTestId('budget-bar-savings');
    expect(budgetBarSavings).toHaveTextContent('Savings');
    expect(budgetBarSavings).toHaveTextContent('$1,516');
    expect(budgetBarSavings).toHaveTextContent('$128 (8%)');

    const budgetBarNeeds = screen.getByTestId('budget-bar-needs');
    expect(budgetBarNeeds).toHaveTextContent('Needs');
    expect(budgetBarNeeds).toHaveTextContent('-$4,053 (107%)');
    expect(budgetBarNeeds).toHaveTextContent('-$3,788');

    const budgetBarWants = screen.getByTestId('budget-bar-wants');
    expect(budgetBarWants).toHaveTextContent('Wants');
    expect(budgetBarWants).toHaveTextContent('-$3,769 (166%)');
    expect(budgetBarWants).toHaveTextContent('-$2,273');

    const budgetBarOutOfBudget = screen.getByTestId('budget-out-of-budget');
    expect(budgetBarOutOfBudget).toHaveTextContent('Out of budget');
    expect(budgetBarOutOfBudget).toHaveTextContent('-$100');

    userEvent.click(screen.getByText('Edit'));
    await waitFor(() => {
      expect(screen.getByText('What is auto-budget?')).toBeVisible();
      expect(screen.getByDisplayValue('Enabled')).toBeChecked();
      expect(screen.getByDisplayValue('Disabled')).not.toBeChecked();
      expect(screen.getByText('Save')).toHaveAttribute('disabled');
      expect(screen.queryByText('Add new')).toBeNull();
    });

    const budgetFieldsetTargetIncome = screen.getByTestId('budget-fieldset-target-income');
    expect(budgetFieldsetTargetIncome).toHaveTextContent('Target income');
    expect(budgetFieldsetTargetIncome).toHaveTextContent('100%');
    expect(screen.getByDisplayValue('+$7,577')).toHaveAttribute('disabled');

    let budgetFieldsetExpenseGroups = screen.getAllByTestId('budget-fieldset-expense-group');
    expect(budgetFieldsetExpenseGroups[0]).toHaveTextContent('Expense group name');
    expect(budgetFieldsetExpenseGroups[0]).toHaveTextContent('Target amount');
    expect(budgetFieldsetExpenseGroups[0]).toHaveTextContent('Categories');
    expect(budgetFieldsetExpenseGroups[0]).toHaveTextContent('50%');
    expect(budgetFieldsetExpenseGroups[1]).toHaveTextContent('30%');
    expect(screen.getByDisplayValue('Needs')).toHaveAttribute('disabled');
    expect(screen.getByDisplayValue('-$3,788')).toHaveAttribute('disabled');
    expect(screen.getByDisplayValue('32')).toHaveAttribute('disabled');
    expect(screen.getByDisplayValue('Wants')).toHaveAttribute('disabled');
    expect(screen.getByDisplayValue('-$2,273')).toHaveAttribute('disabled');
    expect(screen.getByDisplayValue('53')).toHaveAttribute('disabled');

    let budgetFieldsetTargetSavings = screen.getByTestId('budget-fieldset-target-savings');
    expect(budgetFieldsetTargetSavings).toHaveTextContent('Target savings');
    expect(budgetFieldsetTargetSavings).toHaveTextContent('20%');
    expect(screen.getByDisplayValue('$1,516')).toHaveAttribute('disabled');

    userEvent.click(screen.getByText('Transaction categories'));
    expect(screen.getByDisplayValue('Enabled')).toHaveAttribute('disabled');
    expect(screen.getByText('Assigning categories')).toBeVisible();
    expect(screen.getByText('Save')).toHaveAttribute('disabled');
    expect(screen.queryAllByRole('row').length).toEqual(86);

    userEvent.click(screen.getByText('Budget groups'));
    await waitFor(() => {
      userEvent.click(screen.getByDisplayValue('Disabled'));
      expect(screen.getByDisplayValue('Disabled')).toBeChecked();
      expect(screen.getByText('Custom budget')).toBeVisible();
      expect(screen.getByDisplayValue('Enabled')).not.toBeChecked();
      expect(screen.getByText('Save')).not.toHaveAttribute('disabled');
    });
    expect(screen.getByDisplayValue('+$7,577')).not.toHaveAttribute('disabled');
    expect(screen.getByDisplayValue('Needs')).not.toHaveAttribute('disabled');
    expect(screen.getByDisplayValue('-$3,788')).not.toHaveAttribute('disabled');
    expect(screen.getByDisplayValue('Wants')).not.toHaveAttribute('disabled');
    expect(screen.getByDisplayValue('-$2,273')).not.toHaveAttribute('disabled');

    // FIXME: When disabling auto-budget from the Edit page, the expense groups are
    // populated with the data from auto-budget but only in the app and not during the tests.
    userEvent.clear(screen.getByDisplayValue('+$7,577'));
    let expenseGroupTargetAmountInput = screen.getAllByPlaceholderText('$0.00');
    expect(expenseGroupTargetAmountInput.length).toEqual(4);

    userEvent.clear(expenseGroupTargetAmountInput[0]);
    userEvent.type(expenseGroupTargetAmountInput[0], '1000');
    userEvent.clear(expenseGroupTargetAmountInput[1]);
    userEvent.type(expenseGroupTargetAmountInput[1], '500');
    userEvent.clear(expenseGroupTargetAmountInput[2]);
    userEvent.type(expenseGroupTargetAmountInput[2], '400');
    budgetFieldsetExpenseGroups = screen.getAllByTestId('budget-fieldset-expense-group');
    expect(budgetFieldsetExpenseGroups.length).toEqual(2);
    expect(budgetFieldsetExpenseGroups[0]).toHaveTextContent('50%');
    expect(budgetFieldsetExpenseGroups[1]).toHaveTextContent('40%');

    budgetFieldsetTargetSavings = screen.getByTestId('budget-fieldset-target-savings');
    expect(budgetFieldsetTargetSavings).toHaveTextContent('10%');
    expect(screen.getByText('Save')).not.toHaveAttribute('disabled');

    const spyOnSubmit = jest.spyOn(BudgetIpc, 'editBudgetGroups');
    userEvent.click(screen.getByText('Save'));
    await waitFor(() => {
      expect(spyOnSubmit).toHaveBeenCalledTimes(1);
      expect(spyOnSubmit).toMatchSnapshot();
    });

    const removeButtons = screen.getAllByText('Remove');
    expect(removeButtons.length).toEqual(2);

    userEvent.click(removeButtons[0]);
    await waitFor(() => {
      budgetFieldsetExpenseGroups = screen.getAllByTestId('budget-fieldset-expense-group');
      expect(budgetFieldsetExpenseGroups.length).toEqual(1);
      expect(budgetFieldsetExpenseGroups[0]).not.toHaveTextContent('50%');
      expect(budgetFieldsetExpenseGroups[0]).toHaveTextContent('40%');

      budgetFieldsetTargetSavings = screen.getByTestId('budget-fieldset-target-savings');
      expect(budgetFieldsetTargetSavings).not.toHaveTextContent('10%');
      expect(budgetFieldsetTargetSavings).toHaveTextContent('60%');
    });

    userEvent.click(removeButtons[1]);
    await waitFor(() => {
      budgetFieldsetExpenseGroups = screen.queryAllByTestId('budget-fieldset-expense-group');
      expect(budgetFieldsetExpenseGroups.length).toEqual(0);

      budgetFieldsetTargetSavings = screen.getByTestId('budget-fieldset-target-savings');
      expect(budgetFieldsetTargetSavings).not.toHaveTextContent('60%');
      expect(budgetFieldsetTargetSavings).toHaveTextContent('100%');
    });
    expect(screen.getByText('No expenses in budget')).toBeVisible();

    userEvent.click(screen.getByText('Add new'));
    await waitFor(() => {
      budgetFieldsetExpenseGroups = screen.getAllByTestId('budget-fieldset-expense-group');
      expect(budgetFieldsetExpenseGroups.length).toEqual(1);
      expect(budgetFieldsetExpenseGroups[0]).toHaveTextContent('0%');
    });

    expenseGroupTargetAmountInput = screen.getAllByPlaceholderText('$0.00');
    expect(expenseGroupTargetAmountInput.length).toEqual(3);
    expect(screen.getByText('Save')).toHaveAttribute('disabled');

    userEvent.type(expenseGroupTargetAmountInput[1], '1500');
    await waitFor(() => {
      budgetFieldsetExpenseGroups = screen.getAllByTestId('budget-fieldset-expense-group');
      expect(budgetFieldsetExpenseGroups[0]).toHaveTextContent('150%');
    });
    budgetFieldsetTargetSavings = screen.getByTestId('budget-fieldset-target-savings');
    expect(budgetFieldsetTargetSavings).toHaveTextContent('-50%');
    expect(screen.getByText('Save')).toHaveAttribute('disabled');
  });

  // TODO: test('Budget page displays the correct data when auto-budget is disabled', async () => {});
});
