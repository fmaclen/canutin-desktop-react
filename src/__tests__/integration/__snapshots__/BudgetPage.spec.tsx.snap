// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Budget tests Budget page displays the correct data when auto-budget is enabled 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "autoBudgetField": "Disabled",
        "editedBudgets": Array [
          Object {
            "name": "Income",
            "targetAmount": 1000,
            "type": "income",
          },
          Object {
            "categories": Array [
              Object {
                "id": 1,
                "name": "Education",
              },
              Object {
                "id": 2,
                "name": "Studen loan",
              },
              Object {
                "id": 3,
                "name": "Groceries",
              },
              Object {
                "id": 4,
                "name": "Health",
              },
              Object {
                "id": 5,
                "name": "Medical care",
              },
              Object {
                "id": 6,
                "name": "Pharmacies",
              },
              Object {
                "id": 7,
                "name": "Housing",
              },
              Object {
                "id": 8,
                "name": "Mortgage",
              },
              Object {
                "id": 9,
                "name": "Rent",
              },
              Object {
                "id": 10,
                "name": "Institutional",
              },
              Object {
                "id": 11,
                "name": "Insurance",
              },
              Object {
                "id": 12,
                "name": "Government",
              },
              Object {
                "id": 13,
                "name": "Legal",
              },
              Object {
                "id": 14,
                "name": "Taxes",
              },
              Object {
                "id": 15,
                "name": "Kids",
              },
              Object {
                "id": 16,
                "name": "Allowance",
              },
              Object {
                "id": 17,
                "name": "Child care",
              },
              Object {
                "id": 18,
                "name": "Kids supplies",
              },
              Object {
                "id": 19,
                "name": "Pets",
              },
              Object {
                "id": 20,
                "name": "Veterinary",
              },
              Object {
                "id": 21,
                "name": "Transportation",
              },
              Object {
                "id": 22,
                "name": "Automotive",
              },
              Object {
                "id": 23,
                "name": "Gas stations",
              },
              Object {
                "id": 24,
                "name": "Service & parts",
              },
              Object {
                "id": 25,
                "name": "Public transportation",
              },
              Object {
                "id": 26,
                "name": "Buses & trains",
              },
              Object {
                "id": 27,
                "name": "Utilities",
              },
              Object {
                "id": 28,
                "name": "Electricity",
              },
              Object {
                "id": 29,
                "name": "Gas",
              },
              Object {
                "id": 30,
                "name": "Internet & phone",
              },
              Object {
                "id": 31,
                "name": "Water",
              },
              Object {
                "id": 32,
                "name": "Sanitary & waste management",
              },
            ],
            "name": "Needs",
            "targetAmount": "-500",
            "type": "expense",
          },
          Object {
            "categories": Array [
              Object {
                "id": 100,
                "name": "Museums",
              },
              Object {
                "id": 101,
                "name": "Music",
              },
              Object {
                "id": 102,
                "name": "Nightlife",
              },
              Object {
                "id": 103,
                "name": "Sports",
              },
              Object {
                "id": 104,
                "name": "Subscriptions",
              },
              Object {
                "id": 105,
                "name": "Theatres",
              },
              Object {
                "id": 106,
                "name": "Outdoors & parks",
              },
              Object {
                "id": 107,
                "name": "Business & services",
              },
              Object {
                "id": 108,
                "name": "Contractors",
              },
              Object {
                "id": 109,
                "name": "Manufacturing",
              },
              Object {
                "id": 110,
                "name": "Office supplies",
              },
              Object {
                "id": 111,
                "name": "Postal & shipping",
              },
              Object {
                "id": 112,
                "name": "Books & supplies",
              },
              Object {
                "id": 113,
                "name": "Financial & banking",
              },
              Object {
                "id": 114,
                "name": "Cash",
              },
              Object {
                "id": 115,
                "name": "Fees",
              },
              Object {
                "id": 116,
                "name": "Financial services",
              },
              Object {
                "id": 117,
                "name": "Income",
              },
              Object {
                "id": 118,
                "name": "Interest",
              },
              Object {
                "id": 119,
                "name": "Payments",
              },
              Object {
                "id": 120,
                "name": "Transfers",
              },
              Object {
                "id": 121,
                "name": "Withdrawals",
              },
              Object {
                "id": 122,
                "name": "Bars",
              },
              Object {
                "id": 123,
                "name": "Coffee shops",
              },
              Object {
                "id": 124,
                "name": "Food & drink",
              },
              Object {
                "id": 125,
                "name": "Restaurants",
              },
              Object {
                "id": 126,
                "name": "Fitness",
              },
              Object {
                "id": 127,
                "name": "Furnishings",
              },
              Object {
                "id": 128,
                "name": "Home maintenance",
              },
              Object {
                "id": 129,
                "name": "Home improvement",
              },
              Object {
                "id": 130,
                "name": "Home security",
              },
              Object {
                "id": 131,
                "name": "Religious",
              },
              Object {
                "id": 132,
                "name": "Toys",
              },
              Object {
                "id": 133,
                "name": "Personal",
              },
              Object {
                "id": 134,
                "name": "Charity",
              },
              Object {
                "id": 135,
                "name": "Gifts",
              },
              Object {
                "id": 136,
                "name": "Payroll & benefits",
              },
              Object {
                "id": 137,
                "name": "Personal care",
              },
              Object {
                "id": 138,
                "name": "Pet services",
              },
              Object {
                "id": 139,
                "name": "Shops",
              },
              Object {
                "id": 140,
                "name": "Arts & crafts",
              },
              Object {
                "id": 141,
                "name": "Clothing",
              },
              Object {
                "id": 142,
                "name": "Electronics",
              },
              Object {
                "id": 143,
                "name": "Hobbies",
              },
              Object {
                "id": 144,
                "name": "Parking",
              },
              Object {
                "id": 145,
                "name": "Taxi & ride sharing",
              },
              Object {
                "id": 146,
                "name": "Travel",
              },
              Object {
                "id": 147,
                "name": "Vehicle rentals",
              },
              Object {
                "id": 148,
                "name": "Boats & cruises",
              },
              Object {
                "id": 149,
                "name": "Lodging",
              },
              Object {
                "id": 150,
                "name": "Vacation",
              },
              Object {
                "id": 151,
                "name": "Television",
              },
              Object {
                "id": 152,
                "name": "Entertainment & recreation",
              },
            ],
            "name": "Wants",
            "targetAmount": "-400",
            "type": "expense",
          },
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
